search-endpoint.md

```diff
@@ -0,0 +1,397 @@
+        "operator" : "authority",
+        "value" : "1234",
+        "label" : "Smith, Donald"
+      }
+  ],
+  "sort" : {
+    "by" : "dateissued",
+    "order" : "asc"
+  },
+  "page": {
+    	"size": 5,
+    	"totalElements": 14,
+    	"totalPages": 3,
+    	"number": 0
+  },
+  "_embedded" : {
+    "searchResults" : [
+      {
+        "hitHighlights": {
+          "dc.description.abstract" : "This is the <em>very cool</em> abstract of this item",
+          "dc.publisher" : "My <em>very cool</em> publisher",
+        },
+        "_links" : {
+          "dspaceObject" : {
+            "href": "/api/core/items/9f3288b2-f2ad-454f-9f4c-70325646dcee"
+          }
+        },
+        "_embedded" : {
+          "dspaceObject" : {
+            "uuid": "9f3288b2-f2ad-454f-9f4c-70325646dcee",
+            "name": "Test Webpage",
+            "handle": "10673/4"
+          }
+        }
+      },
+      {
+        "hitHighlights": { },
+        "_links" : {
+          "dspaceObject" : {
+            "href": "/api/core/items/ff7ec3a4-0aab-418b-94fc-d0e8189084db"
+          }
+        },
+        "_embedded" : {
+          "dspaceObject" : {
+            "uuid": "ff7ec3a4-0aab-418b-94fc-d0e8189084db",
+            "name": "Test Item with no hit highlights",
+            "handle": "10673/5"
+          }
+        }
+      }
+    ]
+  },
+  "_links": {
+      "first": {
+        "href": "/api/discover/search/<:dso-type>/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&title.notcontains=abcd&author.authority=1234&page=0&size=5"
+      },
+      "self": {
+        "href": "/api/discover/search/<:dso-type>/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&title.notcontains=abcd&author.authority=1234&page=0&size=5"
+      },
+      "next": {
+        "href": "/api/discover/search/<:dso-type>/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&title.notcontains=abcd&author.authority=1234&page=1&size=5"
+      },
+      "last": {
+        "href": "/api/discover/search/<:dso-type>/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&title.notcontains=abcd&author.authority=1234&page=2&size=5"
+      }
+  }
+}
+```
+
+### Matching facet search results
+**/api/discover/search/<:dso-type>/facets**
+
+This endpoint returns a list of facet values and counts based on the current query. The result can be refined using the following parameters:
+* `query`: The discovery search string to will be used to match records.
+* `scope`: UUID of a specific DSpace container (site, community or collection) to which the search has to be limited, e.g. `scope=9076bd16-e69a-48d6-9e41-0238cb40d863`.
+* `<:filter-name>.<:filter-operator>`: Advanced search filter that has to be used to filter the result set. The `filter-name` and `filter-operator` must match a value returned by parent search endpoint (see above). For example `author.authority=5df05073-3be7-410d-8166-e254369e4166` or `title.notcontains=rainbows`.
```

Comment:
You are right, I haven't thinked about repeatable parameters... and it is not appropriate to rely on the parameters order so another option could be
f.author=Deasir,contains&f.author=Tom,notcontains 
this is similar to how we set the order direction in sort parameter.

The f. prefix have two reasons:
- avoid conflicts as we don't know in advance which filter names will be defined (for instance someone could have a filter named scope, configuration, query etc.)
- limit the allowed parameter names that can be sent. It is risky to be open to process any parameter as buggy - too smart - code could bind a "userid" parameter to the internal user identity (of course this is unlike to happen but...)

As automatic binding I mean the "Customizable binding and validation" feature of SpringMVC https://docs.spring.io/spring/docs/current/spring-framework-reference/html/mvc.html#mvc-ann-requestparam